crates/rattler_repodata_gateway/src/fetch/jlap/mod.rs

```diff
@@ -0,0 +1,913 @@
+    pub async fn apply(
+        &self,
+        repo_data_json_path: &Path,
+        hash: Output<Blake2b256>,
+    ) -> Result<(), JLAPError> {
+        // We use the current hash to find which patches we need to apply
+        let current_idx = find_current_patch_index(&self.patches, hash);
+
+        return if let Some(idx) = current_idx {
+            let applicable_patches: Vec<&Patch> =
+                self.patches[idx..self.patches.len()].iter().collect();
+            let new_hash = apply_jlap_patches(&applicable_patches, repo_data_json_path).await?;
+
+            // TODO: This check might be a little redundant considering we have validated our
+            //       checksums by now, but it could be nice to keep here for extra validation.
+            //       We could remove it if performance would benefit.
+            if new_hash != self.footer.latest.unwrap_or_default() {
+                return Err(JLAPError::HashesNotMatching);
+            }
+
+            Ok(())
+        } else {
+            Err(JLAPError::NoHashFound)
+        };
+    }
+
+    /// Returns a new JLAPState based on values in JLAP object
+    ///
+    /// We accept `position` as an argument because it is not derived from the JLAP response.
+    ///
+    /// The `iv` (initialization vector) value is optionally passed in because we may wish
+    /// to override what was initially stored there, which would be the value calculated
+    /// with `validate_checksum`.
+    pub fn get_state(&self, position: u64, iv: Option<Output<Blake2b256>>) -> JLAPState {
+        let iv = match iv {
+            Some(value) => format!("{:x}", value),
+            None => hex::encode(&self.initialization_vector),
+        };
+        JLAPState {
+            pos: position + self.bytes_offset,
+            iv,
+            footer: self.footer.clone(),
+        }
+    }
+
+    /// Validates the checksum present on a JLAP object
+    pub fn validate_checksum(&self) -> Result<Output<Blake2b256>, JLAPError> {
+        let mut iv: Option<Output<Blake2b256>> = None;
+        let mut iv_values: Vec<Output<Blake2bMac256>> = vec![];
+        let end = self.lines.len() - JLAP_FOOTER_OFFSET;
+
+        for line in &self.lines[self.offset..end] {
+            match iv {
+                Some(value) => {
+                    iv = Some(blake2b_256_hash_with_key(line.as_bytes(), &value));
+                }
+                None => {
+                    iv = Some(blake2b_256_hash_with_key(
+                        line.as_bytes(),
+                        &self.initialization_vector,
+                    ));
+                }
+            }
+            iv_values.push(iv.unwrap());
+        }
+
+        if !iv_values.is_empty() {
+            let new_iv = iv_values[iv_values.len() - 1];
+            if new_iv != self.checksum {
+                return Err(JLAPError::ChecksumMismatch);
+            }
+
+            Ok(new_iv)
+        } else {
+            Err(JLAPError::NoPatchesFound)
+        }
```

Comment:
Indexing vectors is a bit unsafe. Although it doesn't result in undefined behavior it does panic if you index incorrectly. To that end I always prefer to use more semantic means:

```suggestion
        if let Some(new_iv) = iv_values.pop() {
            if new_iv != self.checksum {
                return Err(JLAPError::ChecksumMismatch);
            }
            Ok(new_iv)
        } else {
            Err(JLAPError::NoPatchesFound)
        }
```

Reply:
Coming from Python, I miss not being able to do:

```python
something = [1 ,2 ,3]
print(something[-1]) # print the last element
```

😅 

Thanks for the suggested improvement.